### PR TITLE
Add PayatTable features and Pre-auth features

### PR DESCRIPTION
**Pull request by metinavc1**

### CODE REVIEW

File: Library/Model/SPIPayAtTable.h [hunk 1]
```diff
@@ -100,15 +100,16 @@ typedef NS_ENUM(NSUInteger, SPIPaymentType) {
...
 @property (nonatomic) BOOL operatorIdEnabled;
 @property (nonatomic) BOOL splitByAmountEnabled;
 @property (nonatomic) BOOL equalSplitEnabled;
 @property (nonatomic) BOOL tippingEnabled;
 @property (nonatomic) BOOL summaryReportEnabled;
-@property (nonatomic, readonly, copy) NSString *labelPayButton;
-@property (nonatomic, readonly, copy) NSString *labelOperatorId;
-@property (nonatomic, readonly, copy) NSString *labelTableId;
-@property (nonatomic, readonly, copy) NSArray<NSString *> *allowedOperatorIds;
+@property (nonatomic) NSString *labelPayButton;
```
Comment (gouline) [inline Library/Model/SPIPayAtTable.h]: Reason for change?

Comment (metinavc1) [reply Library/Model/SPIPayAtTable.h]: They couldn't be set a value when they were read-only

Comment (gouline) [reply Library/Model/SPIPayAtTable.h]: I was more talking about the removal of `copy`, this will impact the behaviour.

Comment (metinavc1) [reply Library/Model/SPIPayAtTable.h]: ok I got it. Should it be like that `@property (nonatomic, copy) NSString *labelPayButton;`?

**Comment by chrishulbert** [comment]
Hi, I’ll be glad to have a look at this if my contract gets renewed in the new year :)

File: Library/Model/SPIPayAtTable.h (hunk 1 above)
Comment (chrishulbert) [reply Library/Model/SPIPayAtTable.h]: Copy is usually a safe bet for things like strings / arrays where performance isn't a massive deal.

File: Library/Model/SPIPayAtTable.m [hunk 2]
```diff
@@ -19,57 +19,64 @@ @implementation SPIBillStatusResponse
...
     NSError *writeError = nil;
     NSData *jsonData = [NSJSONSerialization dataWithJSONObject:ph options:NSJSONWritingPrettyPrinted error:&writeError];
     return [jsonData base64EncodedStringWithOptions:0];
 }
 
 - (SPIMessage *)toMessage:(NSString *)messageId {
     NSMutableDictionary *data = [[NSMutableDictionary alloc] init];
-    [data setObject:[NSNumber numberWithInteger:BillRetrievalResultSuccess] forKey:@"success"];
+    BOOL success = _result == BillRetrievalResultSuccess;
+    [data setValue:[NSNumber numberWithBool:success] forKey:@"success"];
```
Comment (chrishulbert) [inline Library/Model/SPIPayAtTable.m]: IIRC, this is a shortcut for nsnumber:
```suggestion
    [data setValue:@success forKey:@"success"];
```

File: Library/Model/SPIPayAtTable.m [hunk 3]
```diff
@@ -19,57 +19,64 @@ @implementation SPIBillStatusResponse
...
     }
     if (_tableId.length > 0) {
-        [data setObject:_tableId forKey:@"table_id"];
+        [data setValue:_tableId forKey:@"table_id"];
     }
     if (_result == BillRetrievalResultSuccess) {
-        [data setObject:@"bill_total_amount" forKey:[NSNumber numberWithInteger:_totalAmount]];
-        [data setObject:@"bill_outstanding_amount" forKey:[NSNumber numberWithInteger:_outstandingAmount]];
-        [data setObject:@"bill_payment_history" forKey:[self getBillPaymentHistory]];
+        [data setValue:[NSNumber numberWithInteger:_totalAmount] forKey:@"bill_total_amount"];
```
Comment (chrishulbert) [inline Library/Model/SPIPayAtTable.m]: Should work if you want to remember for next time, no big issue if you dont do this for this version: ```suggestion
        [data setValue:@_totalAmount forKey:@"bill_total_amount"];
```

File: Library/Model/SPIPayAtTable.m [hunk 4]
```diff
@@ -112,35 +119,45 @@ - (instancetype)initWithMessage:(SPIMessage *)message {
...
+    
+    if ([paymentType isEqualToString:cardPaymentType]) {
         _paymentType = SPIPaymentTypeCard;
-    } else if ([paymentType isEqualToString:[SPIBillPayment paymentTypeString:SPIPaymentTypeCash]]) {
+    } else if ([paymentType isEqualToString:cashPaymentType]) {
         _paymentType = SPIPaymentTypeCash;
     }
-    NSDictionary<NSString *,NSObject *> *data = (NSDictionary *)[message.data valueForKey:@"payment_details"];
     
+    NSDictionary<NSString *,NSObject *> *data = (NSDictionary *)[message.data valueForKey:@"payment_details"];
```
Comment (chrishulbert) [inline Library/Model/SPIPayAtTable.m]:
```suggestion
    NSDictionary<NSString *,NSObject *> *data = (NSDictionary *)message.data[@"payment_details"];
```
Up to you if you want to use this shortcut in future

File: Library/Model/SPIPayAtTable.m [hunk 5]
```diff
@@ -149,17 +166,17 @@ @implementation SPIPayAtTableConfig
...
+    [data setValue:[NSNumber numberWithBool:_payAtTableEnabled]  forKey:@"pay_at_table_enabled"];
     [data setValue:[NSNumber numberWithBool:_operatorIdEnabled] forKey:@"operator_id_enabled"];
     [data setValue:[NSNumber numberWithBool:_splitByAmountEnabled] forKey:@"split_by_amount_enabled"];
     [data setValue:[NSNumber numberWithBool:_equalSplitEnabled] forKey:@"equal_split_enabled"];
     [data setValue:[NSNumber numberWithBool:_tippingEnabled] forKey:@"tipping_enabled"];
     [data setValue:[NSNumber numberWithBool:_summaryReportEnabled] forKey:@"summary_report_enabled"];
-    [data setValue:[NSNumber numberWithBool:_labelPayButton] forKey:@"pay_button_label"];
-    [data setValue:[NSNumber numberWithBool:_labelOperatorId] forKey:@"operator_id_label"];
-    [data setValue:[NSNumber numberWithBool:_labelTableId] forKey:@"table_id_label"];
+    [data setValue:[NSString stringWithFormat:@"%@", _labelPayButton] forKey:@"pay_button_label"];
```
Comment (chrishulbert) [inline Library/Model/SPIPayAtTable.m:175]: This (and the next 2 lines) looks peculiar. Are we serialising the descriptions of buttons into JSON? That doesn't seem right. Am i missing something here? Should it rather be something like `setValue:_labelPayButton.text` ?

File: Library/Model/SPIPayAtTable.m [hunk 6]
```diff
@@ -184,6 +201,18 @@ @implementation SPIPayAtTable
 
 - (instancetype)initWithClient:(SPIClient *)spi {
     _spi = spi;
+    _config = [SPIPayAtTableConfig alloc];
```
Comment (chrishulbert) [inline Library/Model/SPIPayAtTable.m]: please call `init` as well :)

File: Library/Model/SPIPreAuth.h [hunk 7]
```diff
@@ -7,11 +7,186 @@
...
+extern NSString *const SPIPreauthPartialCancellationRequestKey;
+extern NSString *const SPIPreauthPartialCancellationResponseKey;
+extern NSString *const SPIPreauthCancellationRequestKey;
+extern NSString *const SPIPreauthCancellationResponseKey;
+extern NSString *const SPIPreauthCompleteRequestKey;
+extern NSString *const SPIPreauthCompleteResponseKey;
 
 @class SPIClient;
 
+typedef void (^SPICompletionTxResult)(SPIInitiateTxResult *result);
```
Comment (chrishulbert) [inline Library/Model/SPIPreAuth.h:30]: Great to see a block typedef :)

File: Library/Model/SPIPreAuth.h [hunk 8]
```diff
@@ -7,11 +7,186 @@
...
+- (instancetype)initWithPosRefId:(NSString *)posRefId;
+
+- (SPIMessage *)toMessage;
+
+@end
+
+@interface SPIAccountVerifyResponse : NSObject
+
+@property (nonatomic, readonly, copy) NSString *posRefId;
+@property (nonatomic, readonly, copy) SPIPurchaseResponse *details;
```
Comment (chrishulbert) [inline Library/Model/SPIPreAuth.h:94]: Normally you just use 'copy' for strings/arrays/dictionaries - you can certainly use it for your own types like this but be careful that the type has NSCopying implemented that's all :)

File: Library/SPIClient.m [hunk 9]
```diff
@@ -103,6 +103,12 @@ - (SPIPayAtTable *)enablePayAtTable {
     return _spiPat;
 }
 
+- (SPIPayAtTable *)disablePayAtTable {
+    _spiPat = [[SPIPayAtTable alloc] initWithClient:self];
+    _spiPat.config.payAtTableEnabled = true;
```
Comment (chrishulbert) [inline Library/SPIClient.m]: should this be false to disable it as per the method name?

**Comment by chrishulbert** [comment]
Ok my review is complete, if you can just work through my comments, let me know how you go, and i'll approve it, nothing major, mostly just things i think might be worth double checking :)

File: Library/Model/SPIPayAtTable.h (hunk 1 above)
Comment (metinavc1) [reply Library/Model/SPIPayAtTable.h]: Added like that @property (nonatomic, copy) NSString *labelPayButton; @chrishulbert

File: Library/Model/SPIPayAtTable.m (hunk 2 above)
Comment (metinavc1) [reply Library/Model/SPIPayAtTable.m]: it gives the error "Unexpected ''@" in program " when I change it like your suggestion @chrishulbert

File: Library/Model/SPIPayAtTable.m (hunk 3 above)
Comment (metinavc1) [reply Library/Model/SPIPayAtTable.m]: it gives the error "Unexpected ''@" in program " when I change it like your suggestion @chrishulbert

File: Library/Model/SPIPayAtTable.m (hunk 2 above)
Comment (chrishulbert) [reply Library/Model/SPIPayAtTable.m]: My ObjC is rusty... perhaps it's `@(success)`. Nevermind, it's just a style suggestion to keep in mind :)

Comment (metinavc1) [reply Library/Model/SPIPayAtTable.m]: Done

File: Library/Model/SPIPayAtTable.m (hunk 3 above)
Comment (metinavc1) [reply Library/Model/SPIPayAtTable.m]: Done

**Review by chrishulbert — CHANGES_REQUESTED** [review]
Just a couple little changes cheers 🍻

File: Library/Model/SPIPayAtTable.m (hunk 4 above)
Comment (metinavc1) [reply Library/Model/SPIPayAtTable.m]: ok i got it

File: Library/Model/SPIPayAtTable.m (hunk 6 above)
Comment (metinavc1) [reply Library/Model/SPIPayAtTable.m]: Done

File: Library/Model/SPIPayAtTable.m (hunk 5 above)
Comment (metinavc1) [reply Library/Model/SPIPayAtTable.m:175]: I checked it that doesn't work.

File: Library/Model/SPIPreAuth.h (hunk 7 above)
Comment (metinavc1) [reply Library/Model/SPIPreAuth.h:30]: 👍 thanks :)

File: Library/Model/SPIPreAuth.h (hunk 8 above)
Comment (metinavc1) [reply Library/Model/SPIPreAuth.h:94]: ok i understand :)

File: Library/SPIClient.m (hunk 9 above)
Comment (metinavc1) [reply Library/SPIClient.m]: yes right thanks :)

File: Library/Model/SPIPayAtTable.m (hunk 5 above)
Comment (chrishulbert) [reply Library/Model/SPIPayAtTable.m:175]: So i’ve looked in the header, labelPayButton isn’t actually a UIButton which i assumed, it’s just a NSString
In which case there’s no point doing the stringWithFormat, you can just set it as the dictionary value directly
but there’s no harm either, it’s just an odd thing, so i’m happy to approve

**Review by chrishulbert — APPROVED** [review]

**Pull request merged**